### PR TITLE
Support multiple mappers for each class

### LINKED ISSUE #70: Support multiple mappers per input class type
In `KamelConfig` it would need to support multiple mappers per KClass

`public val mappers: Map<KClass<*>, Mapper<Any, Any>>` -> `public val mappers: Map<KClass<*>, List<Mapper<Any, Any>>>` 

and would need to be handled here:

[https://github.com/Kamel-Media/Kamel/blob/c83ef4ae6ad05bd58a00a51352bda7068057cd02[…]l-core/src/commonMain/kotlin/io/kamel/core/utils/ConfigUtils.kt](https://github.com/Kamel-Media/Kamel/blob/c83ef4ae6ad05bd58a00a51352bda7068057cd02/kamel-core/src/commonMain/kotlin/io/kamel/core/utils/ConfigUtils.kt#L10-L12)

### PR DESCRIPTION
**Pull request by iruizmar**
Closes #70

### CODE REVIEW

**Review by luca992 — CHANGES_REQUESTED** [review]
Looks good besides that

File: kamel-core/src/commonTest/kotlin/io/kamel/core/config/KamelConfigBuilderTest.kt
Comment (luca992) [inline kamel-core/src/commonTest/kotlin/io/kamel/core/config/KamelConfigBuilderTest.kt:1]: You should just be able to use flatten

Comment (iruizmar) [reply kamel-core/src/commonTest/kotlin/io/kamel/core/config/KamelConfigBuilderTest.kt:1]: @luca992 I can change it, but this way it's also testing that the mappers are added to the correct key. Is it over-testing in your opinion?

Comment (luca992) [reply kamel-core/src/commonTest/kotlin/io/kamel/core/config/KamelConfigBuilderTest.kt:1]: hmm true. technically that's tested with lines above in the single mapper tests at least. But I guess its good to be more explicit. could you just use getOrDefault to return an empty list to remove the assert not null then?

File: kamel-core/src/commonMain/kotlin/io/kamel/core/config/KamelConfigBuilder.kt [hunk 1]
```diff
@@ -48,7 +48,8 @@ public class KamelConfigBuilder {
     }
 
     public fun <I : Any, O : Any> mapper(mapper: Mapper<I, O>) {
-        mappers[mapper.inputKClass] = mapper as Mapper<Any, Any>
+        mappers[mapper.inputKClass] =
+            (mappers[mapper.inputKClass] ?: emptyList()) + (mapper as Mapper<Any, Any>)
```
Comment (luca992) [inline kamel-core/src/commonMain/kotlin/io/kamel/core/config/KamelConfigBuilder.kt:52]: actually you can use `getOrPut` here too

File: kamel-core/src/commonMain/kotlin/io/kamel/core/config/KamelConfigBuilder.kt [hunk 2]
```diff
@@ -128,7 +130,7 @@ public fun KamelConfigBuilder.takeFrom(config: KamelConfig): KamelConfigBuilder
     svgCacheSize = config.svgCache.maxSize
     config.fetchers.forEach { fetcher(it) }
     config.decoders.forEach { decoder(it) }
-    config.mappers.values.forEach { mapper(it) }
+    config.mappers.values.forEach { mappers -> mappers.forEach { mapper(it) } }
```
Comment (luca992) [inline kamel-core/src/commonMain/kotlin/io/kamel/core/config/KamelConfigBuilder.kt:133]: And use flatten here

File: kamel-core/src/commonTest/kotlin/io/kamel/core/config/KamelConfigBuilderTest.kt
Comment (iruizmar) [reply kamel-core/src/commonTest/kotlin/io/kamel/core/config/KamelConfigBuilderTest.kt:1]: @luca992 I tried your syntax and found a less verbose one, LMK if you like it.

File: kamel-core/src/commonMain/kotlin/io/kamel/core/utils/ConfigUtils.kt [hunk 3]
```diff
@@ -8,7 +8,11 @@ import io.kamel.core.fetcher.Fetcher
 import kotlin.reflect.KClass
 
 internal fun KamelConfig.mapInput(input: Any, inputKClass: KClass<*>): Any {
-    val output = mappers[inputKClass]?.map(input)
+
+    val output = mappers[inputKClass]
+        ?.firstOrNull { mapper -> with(mapper) { input.isSupported } }
```
Comment (luca992) [inline kamel-core/src/commonMain/kotlin/io/kamel/core/utils/ConfigUtils.kt:13]: and sorry, use `lastOrNull` to match the behavior of fetchers.

**Comment by luca992** [comment]
Sorry, I looked at it again, and made more change requests

**Comment by iruizmar** [comment]
@luca992 Sure! I think you forgot to send them, though.

File: kamel-core/src/commonTest/kotlin/io/kamel/core/config/KamelConfigBuilderTest.kt
Comment (luca992) [reply kamel-core/src/commonTest/kotlin/io/kamel/core/config/KamelConfigBuilderTest.kt:1]: Looks good

**Comment by luca992** [comment]
Alright, once again, sorry I'm thinking there might be a better way instead of using:

`val mappers: MutableMap<KClass<*>, List<Mapper<Any, Any>>> = mutableMapOf()`

to use:

`val mappers: MutableMap<Pair<KClass<*>, KClass<*>>, Mapper<Any, Any>> = mutableMapOf()`

what do you think?

**Comment by luca992** [comment]
> > Sorry, I looked at it again, and made more change requests
> 
> @luca992 Sure! I think you forgot to send them, though.

I think you should be able to see them above. Lmk if you do not

**Comment by iruizmar** [comment]
> > > Sorry, I looked at it again, and made more change requests
> > 
> > 
> > @luca992 Sure! I think you forgot to send them, though.
> 
> I think you should be able to see them above. Lmk if you do not

Can't see anything, mate! Maybe you started a review and didn't send it?

**Review by luca992 — CHANGES_REQUESTED** [review]

**Comment by luca992** [comment]
> > > > Sorry, I looked at it again, and made more change requests
> > > 
> > > 
> > > @luca992 Sure! I think you forgot to send them, though.
> > 
> > 
> > I think you should be able to see them above. Lmk if you do not
> 
> Can't see anything, mate! Maybe you started a review and didn't send it?

Sorry, thought I added them to the old review. I guess not.

File: kamel-core/src/commonMain/kotlin/io/kamel/core/config/KamelConfigBuilder.kt (hunk 1 above)
Comment (iruizmar) [reply kamel-core/src/commonMain/kotlin/io/kamel/core/config/KamelConfigBuilder.kt:52]: Had to make the list of mappers a `MutableList`, though.

**Comment by iruizmar** [comment]
> > > > > Sorry, I looked at it again, and made more change requests
> > > > 
> > > > 
> > > > @luca992 Sure! I think you forgot to send them, though.
> > > 
> > > 
> > > I think you should be able to see them above. Lmk if you do not
> > 
> > 
> > Can't see anything, mate! Maybe you started a review and didn't send it?
> 
> Sorry, thought I added them to the old review. I guess not.

@luca992 all good. I addressed them.

**Review by luca992 — APPROVED** [review]

**Comment by luca992** [comment]
Sweet looks good. Thanks!

**Pull request merged**